README.md
# dnstweak

**Quick local DNS spoofing tool**

`dnstweak` is a program that allows you to run something like:

    $ sudo ./dnstweak foo.example.com=10.0.0.1,10.0.0.2

So that `foo.example.com` appears to resolve to `10.0.0.1` and `10.0.0.2`.

It runs a DNS server that proxies all other lookups to the previous resolver from
`/etc/resolv.conf`, and it inserts itself as the new resolver in
`/etc/resolv.conf`.

Lookups for `foo.example.com` will get the 2 given A records
(`10.0.0.1` and `10.0.0.2`) in a random order. It also logs requests and
responses to stdout. Once you stop it with Ctrl-C it will put the previous
`/etc/resolv.conf` contents back in place.

See the [Releases](https://github.com/jes/dnstweak/releases/) page for Linux
binaries, or build it yourself with `go build`.

## Features

`dnstweak` can:

 - forge responses to select DNS requests configured on the command line
 - automatically take over all system DNS queries by inserting itself into `/etc/resolv.conf`
 - lookup client addresses in `procfs` to find out which process a request came from

## Why?

Sometimes you want to test how a piece of software responds to different DNS
records without actually changing the real DNS records.

## Example session

Open 2 terminal windows. We'll run `dnstweak` in the first and `ping` in
the second.

First terminal:

    $ sudo ./dnstweak example.com=127.0.0.1
    2023/07/11 20:41:35 dnstweak starts
    2023/07/11 20:41:35 listening on 127.0.0.1:53
    2023/07/11 20:41:35 using 127.0.0.53:53 as upstream resolver

`dnstweak` stays running. It has inserted itself as the local resolver by
modifying `/etc/resolv.conf`.

If you run `dnstweak` as a non-root user, it will still work as a DNS server,
but it won't be able to listen on port 53 or insert itself into `/etc/resolv.conf`.

In the other terminal, start a `ping` to `example.com`:

    $ ping example.com

Now we get some log output from `dnstweak` in our first terminal:

    2023/07/11 20:41:37 127.0.0.1:48439 (ping/9975): A example.com: 127.0.0.1 (overridden)

A request came from `127.0.0.1:48439`, which is a process called `ping` with
PID 9975.
It was an A lookup for `example.com`, and we returned `127.0.0.1`. Back in the other terminal:

    PING example.com (127.0.0.1) 56(84) bytes of data.
    64 bytes from localhost (127.0.0.1): icmp_seq=1 ttl=64 time=0.020 ms
    64 bytes from localhost (127.0.0.1): icmp_seq=2 ttl=64 time=0.037 ms

Great success.

## Installation

Either get a binary from the [Releases](https://github.com/jes/dnstweak/releases/) page, and:

    $ sudo cp dnstweak.x86_64 /usr/bin/dnstweak
    $ sudo chmod +x /usr/bin/dnstweak

Or build it yourself (see below).

## Usage

    dnstweak v0.2

    usage: dnstweak [options] SPEC...

    options:
      -listen string
            listen address (IP:PORT or just PORT) (default: see below)
      -no-proc
            disable discovering the client process by looking in /proc
      -no-resolvconf
            disable automatic update of /etc/resolv.conf
      -upstream string
            upstream DNS server (IP:PORT or just IP) (default: see below)

    In the absence of -listen, dnstweak will first try to listen on any loopback IP
    address (127.0.0.0/24) on port 53, and failing that use a random port number on
    127.0.0.1.

    In the absence of -upstream, dnstweak will take the first nameserver configured
    in /etc/resolv.conf.

    Each SPEC is a hostname, followed by an "=" sign, followed by a
    comma-separated list of 1 or more IP addresses (for example
    "example.com=127.0.0.1").

    dnstweak is a program by James Stanley. You can email me at
    james@incoherency.co.uk or read my blog at https://incoherency.co.uk/

## Build

`dnstweak` is written in go. You can build it with:

    $ go build

And then run:

    $ ./dnstweak -help

Run the tests with:

    $ go test

## Future

In the future, maybe `dnstweak` will gain options to:

 - listen on IPv6
 - populate the DNS override map from `/etc/hosts`
 - override responses to AAAA, CNAME, PTR, SRV requests
 - make fake NXDOMAIN responses
 - drop requests
 - take a zonefile in a BIND-ish format instead of the made-up command-line format

## What happens to `resolv.conf` if `dnstweak` crashes?

It will look something like this:

    # created by dnstweak
    nameserver 127.0.0.1
    search lan
    #dnstweak## This file is managed by man:systemd-resolved(8). Do not edit.
    #dnstweak##
        [...]
    #dnstweak#
    #dnstweak#nameserver 127.0.0.53
    #dnstweak#options edns0 trust-ad
    #dnstweak#search lan

The lines prefixed with `#dnstweak#` are the lines that were there before `dnstweak`
took over. Delete the uncommented lines at the top, and remove the "`#dnstweak#`"
prefix from the rest, and you'll have your old `resolv.conf` back.

## Other tools like this

https://github.com/iphelix/dnschef

https://github.com/marekjelen/dnshack

## Contact

James Stanley

https://incoherency.co.uk

james@incoherency.co.uk


CHANGELOG.md
# Changelog

## v0.2 - 2022.07-14
 - Fix some cases where dnstweak can crash without restoring `resolv.conf`
 - Leave a commented copy of the original `resolv.conf` to aid restoring it
 - Better log output
 - Stop writing port numbers in the `nameserver` line in `resolv.conf`
 - Automated testing

## v0.1 - 2022-07-12
 - Initial release


dnstweak_test.go
package main

import (
	"testing"
	"time"

	"github.com/miekg/dns"
)

func Test(t *testing.T) {
	// setup an upstream server answering for "upstream.example.com"
	upstreamHost := "upstream.example.com"
	upstreamIp := "1.2.3.4"
	upstreamAddress := "127.0.0.1:12345" // TODO: let linux pick the port
	upstream := makeDnsTweak("10.10.10.10:53", upstreamHost, upstreamIp)

	// setup a local server answering for "local.example.com"
	localHost := "local.example.com"
	localIp := "5.6.7.8"
	localAddress := "127.0.0.1:12346" // TODO: let linux pick the port
	local := makeDnsTweak(upstreamAddress, localHost, localIp)

	// no answers before we start the servers running
	checkNoAnswer(upstreamAddress, upstreamHost, "before starting", t)
	checkNoAnswer(localAddress, localHost, "before starting", t)

	// start the servers
	go runServer(upstream, upstreamAddress, t)
	go runServer(local, localAddress, t)

	// TODO: time is not a synchronisation primitive (take a channel or something to say when they started?)
	time.Sleep(1 * time.Second)

	// sensible answers after they start running
	checkAnswer(upstreamAddress, upstreamHost, upstreamIp, "own override", t)
	checkAnswer(localAddress, localHost, localIp, "own override", t)
	checkNoAnswer(upstreamAddress, localHost, "upstream doesn't answer for local", t)
	checkAnswer(localAddress, upstreamHost, upstreamIp, "local forwards to upstream", t)

	upstream.Finish()

	// no more upstream after upstream has finished
	checkNoAnswer(upstreamAddress, upstreamHost, "upstream has finished", t)
	checkAnswer(localAddress, localHost, localIp, "local still works after upstream has finished", t)
	checkNoAnswer(localAddress, upstreamHost, "local doesn't get a response from upstream after upstream has finished", t)

	local.Finish()

	// no more local after local has finished
	checkNoAnswer(localAddress, localHost, "local has finished", t)
}

func runServer(d *DnsTweak, listenAddress string, t *testing.T) {
	err := d.ListenAndServe(listenAddress)
	if err != nil {
		t.Errorf("%v", err)
	}
}

func dnsQuery(addr string, host string) (*dns.Msg, error) {
	req := &dns.Msg{
		MsgHdr: dns.MsgHdr{
			Opcode: dns.OpcodeQuery,
		},
		Question: make([]dns.Question, 1),
	}
	req.Question[0] = dns.Question{Name: dns.Fqdn(host), Qtype: dns.TypeA, Qclass: uint16(dns.ClassINET)}
	c := new(dns.Client)
	r, _, err := c.Exchange(req, addr)
	return r, err
}

func checkAnswer(addr string, host string, ip string, test string, t *testing.T) {
	r, err := dnsQuery(addr, host)
	if err != nil {
		t.Errorf("%s: (@%s: %s): expected success, got error: %v", test, addr, host, err)
		return
	}
	if len(r.Answer) != 1 {
		t.Errorf("%s: (@%s: %s): expected 1 answer, got %d: %v", test, addr, host, len(r.Answer), r.Answer)
		return
	}
	switch r.Answer[0].(type) {
	case *dns.A:
		ans := r.Answer[0].(*dns.A)
		if ans.A.String() != ip {
			t.Errorf("%s: (@%s: %s): expected %s, got %s: %v", test, addr, host, ip, ans.A, r.Answer)
		}
	default:
		t.Errorf("%s (@%s: %s): expected answer to be A record: %v", test, addr, host, r.Answer)
	}
}

func checkNoAnswer(addr string, host string, test string, t *testing.T) {
	r, err := dnsQuery(addr, host)
	if err == nil {
		t.Errorf("%s (@%s: %s): expected failure, got response: %v", test, addr, host, r.Answer)
	}
}

func makeDnsTweak(upstream string, host string, ip string) *DnsTweak {
	d := DnsTweak{
		Upstream:             upstream,
		SpliceIntoResolvConf: false,
		LookInProc:           false,
	}
	d.AddOverrideSpec(host + "=" + ip)
	return &d
}


proc.go
package main

import (
	"bufio"
	"fmt"
	"io/fs"
	"log"
	"net"
	"os"
	"strconv"
	"strings"
	"syscall"
)

func FindProcess(addr *net.UDPAddr) string {
	// look in /proc/net/udp to find the inode of this address, then look
	// through all of the /proc/$pid/fd/* links to find which one corresponds
	// to this socket, and then look in /proc/$pid/cmdline to find the name
	// of the process

	inode, ok := FindInode(addr)
	if !ok {
		return ""
	}

	entries, err := os.ReadDir("/proc")
	if err != nil {
		log.Printf("%v\n", err)
		return ""
	}

	for _, e := range entries {
		if e.Name() == "1" {
			continue
		}
		if isPID(e.Name()) {
			pid, _ := strconv.Atoi(e.Name())
			if HasInode(pid, inode) {
				// XXX: race window here
				return ProcessName(pid)
			}
		}
	}

	return ""
}

func FindInode(addr *net.UDPAddr) (uint64, bool) {
	// look in /proc/net/udp to find the inode for this address

	f, err := os.Open("/proc/net/udp")
	if err != nil {
		log.Printf("%v\n", err)
		return 0, false
	}
	defer f.Close()

	addrIpInt := int64(addr.IP[0]) + int64(addr.IP[1])*256 + int64(addr.IP[2])*256*256 + int64(addr.IP[3])*256*256*256

	scanner := bufio.NewScanner(f)
	_ = scanner.Scan() // skip first line
	for scanner.Scan() {
		// e.g. " 9777: 0100007F:0035 00000000:0000 07 00000000:00000000 00:00000000 00000000     0        0 2969713 2 ffff9ce6fd1cb180 0"
		line := scanner.Text()
		line = strings.TrimLeft(line, " \t")
		cols := strings.Fields(line)
		ipHex, portHex, ok := strings.Cut(cols[1], ":")
		if !ok {
			log.Printf("/proc/net/udp does not appear to be in the expected format (\"%s\" should look like \"0100007F:9A2A\")\n", cols[1])
			continue
		}
		ipInt, err := strconv.ParseInt(ipHex, 16, 64)
		if err != nil {
			log.Printf("%v\n", err)
			continue
		}
		port, err := strconv.ParseInt(portHex, 16, 0)
		if err != nil {
			log.Printf("%v\n", err)
			continue
		}
		if int(port) == addr.Port && ipInt == addrIpInt {
			inode, err := strconv.ParseUint(cols[9], 10, 64)
			if err != nil {
				log.Printf("/proc/net/udp does not appear to be in the expected format (\"%s\" should be a number)\n", cols[9])
			} else {
				return inode, true
			}
		}
	}

	return 0, false
}

func isPID(name string) bool {
	if len(name) == 0 {
		return false
	}
	for _, c := range name {
		if c < '0' || c > '9' {
			return false
		}
	}
	return true
}

func HasInode(pid int, inode uint64) bool {
	// look under /proc/$pid/fd/ to find a link to the socket inode

	fds, err := os.ReadDir(fmt.Sprintf("/proc/%d/fd", pid))
	if err != nil {
		log.Printf("%v\n", err)
		return false
	}

	for _, fd := range fds {
		fdPath := fmt.Sprintf("/proc/%d/fd/%s", pid, fd.Name())
		stat, err := os.Stat(fdPath)
		if err != nil {
			switch err.(type) {
			case *fs.PathError:
				// this is common, don't bother printing anything
				break
			default:
				log.Printf("%v\n", err)
			}
			continue
		}
		data := stat.Sys()
		switch data.(type) {
		case *syscall.Stat_t:
			s := data.(*syscall.Stat_t)
			if s.Ino == inode {
				return true
			}
		}
	}
	return false
}

func ProcessName(pid int) string {
	// look in /proc/$pid/cmdline

	cmdlineBytes, err := os.ReadFile(fmt.Sprintf("/proc/%d/cmdline", pid))
	if err != nil {
		log.Printf("%v\n", err)
		return fmt.Sprintf("<unknown>/%d", pid)
	}

	cmdline := string(cmdlineBytes[:])
	prog, _, found := strings.Cut(cmdline, "\x00")
	if !found {
		return fmt.Sprintf("<unknown>/%d", pid)
	}

	pathParts := strings.Split(prog, "/")
	return fmt.Sprintf("%s/%d", pathParts[len(pathParts)-1], pid)
}


dnstweak.go
package main

import (
	"fmt"
	"log"
	"math/rand"
	"net"
	"os"
	"os/signal"
	"strings"
	"syscall"

	"github.com/miekg/dns"
)

type DnsTweak struct {
	Override             map[string][]net.IP
	Upstream             string
	SpliceIntoResolvConf bool
	LookInProc           bool
	OldResolvConf        string
	Server               *dns.Server
	PleaseExit           bool
}

func A_record(name string, ipaddr net.IP) *dns.A {
	return &dns.A{
		Hdr: dns.RR_Header{
			Name:   name,
			Rrtype: dns.TypeA,
			Class:  dns.ClassINET,
			Ttl:    60,
		},
		A: ipaddr,
	}
}

func (d *DnsTweak) ServeDNS(w dns.ResponseWriter, req *dns.Msg) {
	clientProcess := ""

	if d.LookInProc {
		switch w.RemoteAddr().(type) {
		case *net.UDPAddr:
			u := w.RemoteAddr().(*net.UDPAddr)
			clientProcess = FindProcess(u)
		}
	}

	if len(req.Question) != 1 {
		if clientProcess == "" {
			log.Printf("%v: unsupported question set length: %d (expected 1): %v\n", w.RemoteAddr(), len(req.Question), req.Question)
		} else {
			log.Printf("%v (%s): unsupported question set length: %d (expected 1): %v\n", w.RemoteAddr(), clientProcess, len(req.Question), req.Question)
		}
		resp := d.PassThrough(req)
		if resp != nil {
			w.WriteMsg(resp)
		}
		return
	}

	resp, overridden := d.Response(req)
	if !overridden {
		resp = d.PassThrough(req)
	}

	d.Log(w, req, resp, overridden, clientProcess)
	if resp != nil {
		w.WriteMsg(resp)
	}
}

func (d *DnsTweak) Log(w dns.ResponseWriter, req *dns.Msg, resp *dns.Msg, overridden bool, client string) {
	qtype := dns.Type(req.Question[0].Qtype)
	name := req.Question[0].Name

	// strip trailing dot from FQDN
	if name[len(name)-1] == '.' {
		name = name[:len(name)-1]
	}

	// format question
	var line string
	if client == "" {
		line = fmt.Sprintf("%v: %v %s: ", w.RemoteAddr(), qtype, name)
	} else {
		line = fmt.Sprintf("%v (%s): %v %s: ", w.RemoteAddr(), client, qtype, name)
	}

	if resp != nil {
		switch req.Question[0].Qtype {
		case dns.TypeA, dns.TypeAAAA, dns.TypePTR:
			ips := make([]string, 0)
			for _, answer := range resp.Answer {
				switch answer.(type) {
				case *dns.A:
					ips = append(ips, answer.(*dns.A).A.String())
				case *dns.AAAA:
					ips = append(ips, answer.(*dns.AAAA).AAAA.String())
				case *dns.CNAME:
					ips = append(ips, strings.TrimSuffix(answer.(*dns.CNAME).Target, "."))
				case *dns.PTR:
					ips = append(ips, strings.TrimSuffix(answer.(*dns.PTR).Ptr, "."))
				default:
					ips = append(ips, answer.String())
				}
			}
			line += strings.Join(ips, ",")
		default:
			// TODO: SRV, more?
			line += fmt.Sprintf("%v", resp.Answer)
		}
	} else {
		line += "(error)"
	}
	if overridden {
		line += " (overridden)"
	}
	log.Printf("%s\n", line)
}

func (d *DnsTweak) Response(req *dns.Msg) (*dns.Msg, bool) {
	qtype := req.Question[0].Qtype
	name := req.Question[0].Name

	if qtype != dns.TypeA {
		return nil, false
	}

	override, exists := d.Override[name]
	if !exists {
		return nil, false
	}

	resp := new(dns.Msg)
	resp.SetReply(req)
	for _, ipaddr := range override {
		resp.Answer = append(resp.Answer, A_record(name, ipaddr))
	}
	rand.Shuffle(len(resp.Answer), func(i, j int) {
		resp.Answer[i], resp.Answer[j] = resp.Answer[j], resp.Answer[i]
	})

	return resp, true
}

func (d *DnsTweak) PassThrough(req *dns.Msg) *dns.Msg {
	c := new(dns.Client)

	resp, _, err := c.Exchange(req, d.Upstream)
	if err != nil {
		log.Printf("%v", err)
		return nil
	}

	return resp
}

func (d *DnsTweak) SetupResolvConf() {
	oldContent, resolver, err := UpdateResolvConf(d.Server.PacketConn.LocalAddr().String())
	if err != nil {
		log.Printf("%v (do you need to be root?)\n", err)
	}
	if resolver != "" {
		if d.Upstream == "" {
			log.Printf("using %s as upstream resolver\n", resolver)
			d.Upstream = resolver
		}
	}
	if oldContent != "" {
		d.OldResolvConf = oldContent
		go d.HandleSignals()
	}
}

func (d *DnsTweak) HandleSignals() {
	c := make(chan os.Signal, 1)
	signal.Notify(c, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGHUP)

	sig := <-c
	log.Printf("received signal: %v\n", sig)
	d.PleaseExit = true
	d.Finish()
}

func (d *DnsTweak) Finish() {
	if d.SpliceIntoResolvConf && d.OldResolvConf != "" {
		err := RestoreResolvConf(d.OldResolvConf)
		if err != nil {
			log.Printf("%v", err)
		}
	}
	if d.Server != nil {
		d.Server.Shutdown()
	}
}

func (d *DnsTweak) ListenAndServe(addr string) error {
	log.Println("dnstweak starts")

	addrs := make([]string, 0)

	if addr != "" {
		addrs = append(addrs, addr)
	} else {
		for i := 1; i < 256; i++ {
			a := fmt.Sprintf("127.0.0.%d:53", i)
			addrs = append(addrs, a)
		}
		addrs = append(addrs, "127.0.0.1:0")
	}

	var err error
	for _, a := range addrs {
		d.Server = &dns.Server{
			Addr:    a,
			Net:     "udp",
			Handler: d,
		}
		d.Server.NotifyStartedFunc = func() {
			log.Printf("listening on %v\n", d.Server.PacketConn.LocalAddr())
			if d.SpliceIntoResolvConf {
				d.SetupResolvConf()
			}
		}

		err = d.Server.ListenAndServe()
		if err != nil {
			log.Printf("%v\n", err)
		}
		if d.PleaseExit {
			os.Exit(0)
		}
		d.Finish() // just in case resolv.conf needs restoring
	}
	return err
}

// a spec should be like foo.example.com=1.2.3.4,5.6.7.8
func (d *DnsTweak) AddOverrideSpec(spec string) {
	if d.Override == nil {
		d.Override = make(map[string][]net.IP)
	}

	host, ips_csv, found := strings.Cut(spec, "=")
	if !found {
		log.Fatalf("spec '%s' does not contain '='\n", spec)
	}

	// append trailing dot for FQDN
	if host[len(host)-1] != '.' {
		host = host + "."
	}
	if _, exists := d.Override[host]; !exists {
		d.Override[host] = make([]net.IP, 0)
	}

	// add each of the IP addresses
	ips := strings.Split(ips_csv, ",")
	for _, ipstr := range ips {
		ip := net.ParseIP(ipstr)
		if ip == nil {
			log.Fatalf("can't parse ip address '%s'", ipstr)
		}
		d.Override[host] = append(d.Override[host], ip)
	}
}


resolvconf.go
package main

import (
	"bufio"
	"os"
	"strings"
)

// change /etc/resolv.conf so that it uses the given resolver;
// success: return (old content, upstream resolver, nil)
// failed to write: return (old content, upstream resolver, err)
// failed to read: return ("", "", err)
func UpdateResolvConf(resolver string) (string, string, error) {
	f, err := os.Open("/etc/resolv.conf")
	if err != nil {
		return "", "", err
	}
	defer f.Close()

	newContent := "# created by dnstweak\n"
	if strings.Contains(resolver, ":") {
		// strip port number, if any
		resolver, _, _ = strings.Cut(resolver, ":")
	}
	newContent += "nameserver " + resolver + "\n"

	oldResolver := ""
	oldLines := make([]string, 0)
	scanner := bufio.NewScanner(f)
	seenDnsTweak := false
	for scanner.Scan() {
		line := scanner.Text()
		oldLines = append(oldLines, line)
		line = strings.TrimLeft(line, " ")
		if oldResolver == "" && strings.HasPrefix(line, "nameserver ") {
			oldResolver = strings.TrimPrefix(line, "nameserver ") + ":53"
		}

		if strings.HasPrefix(line, "search ") {
			newContent += line + "\n"
		}

		if strings.HasPrefix(line, "#dnstweak#") {
			seenDnsTweak = true
		}
	}

	if seenDnsTweak {
		// if the old content was from dnstweak output, then strip it down to just
		// what was there before dnstweak wrote it
		oldLines2 := make([]string, 0)
		for _, line := range oldLines {
			if strings.HasPrefix(line, "#dnstweak#") {
				oldLines2 = append(oldLines2, strings.TrimPrefix(line, "#dnstweak#"))
			}
		}
		oldLines = oldLines2
	}

	oldContent := strings.Join(oldLines, "\n") + "\n"

	// append the old content to the new content, with "#dnstweak#", so that it can be restored later
	for _, line := range oldLines {
		newContent += "#dnstweak#" + line + "\n"
	}

	err = os.WriteFile("/etc/resolv.conf", []byte(newContent), 0644)
	return oldContent, oldResolver, err
}

func RestoreResolvConf(contents string) error {
	err := os.WriteFile("/etc/resolv.conf", []byte(contents), 0644)
	return err
}


main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"strings"
)

var VERSION = "v0.2"

func main() {
	listen := flag.String("listen", "", "listen address (IP:PORT or just PORT) (default: see below)")
	upstream := flag.String("upstream", "", "upstream DNS server (IP:PORT or just IP) (default: see below)")
	noResolvConf := flag.Bool("no-resolvconf", false, "disable automatic update of /etc/resolv.conf")
	noProc := flag.Bool("no-proc", false, "disable discovering the client process by looking in /proc")
	flag.Usage = func() {
		out := flag.CommandLine.Output()
		fmt.Fprintf(out, "dnstweak %s\n\n", VERSION)
		fmt.Fprintf(out, "usage: dnstweak [options] SPEC...\n\noptions:\n")
		flag.PrintDefaults()
		fmt.Fprintf(out, "\n")
		fmt.Fprintf(out, "In the absence of -listen, dnstweak will first try to listen on any loopback IP\naddress (127.0.0.0/24) on port 53, and failing that use a random port number on\n127.0.0.1.\n\n")
		fmt.Fprintf(out, "In the absence of -upstream, dnstweak will take the first nameserver configured\nin /etc/resolv.conf.\n\n")
		fmt.Fprintf(out, "Each SPEC is a hostname, followed by an \"=\" sign, followed by a\ncomma-separated list of 1 or more IP addresses (for example\n\"example.com=127.0.0.1\").\n\n")
		fmt.Fprintf(out, "dnstweak is a program by James Stanley. You can email me at\njames@incoherency.co.uk or read my blog at https://incoherency.co.uk/\n")
	}
	flag.Parse()

	listenAddress := *listen
	if listenAddress != "" && !strings.Contains(listenAddress, ":") {
		listenAddress = "127.0.0.1:" + listenAddress
	}

	upstreamAddress := *upstream
	if upstreamAddress != "" && !strings.Contains(upstreamAddress, ":") {
		upstreamAddress = upstreamAddress + ":53"
	}

	dnstweak := DnsTweak{
		Upstream:             upstreamAddress,
		SpliceIntoResolvConf: !*noResolvConf,
		LookInProc:           !*noProc,
	}
	for _, spec := range flag.Args() {
		dnstweak.AddOverrideSpec(spec)
	}
	err := dnstweak.ListenAndServe(listenAddress)
	if err != nil {
		log.Fatal(err)
	}
}
